TLM/TLM/Manager/Impl/LaneConnectionManagerData/ConnectionDataBase.cs

```diff
@@ -0,0 +1,187 @@
+namespace TrafficManager.Manager.Impl.LaneConnectionManagerData {
+    using CSUtil.Commons;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using TrafficManager.Util;
+    using TrafficManager.Util.Extensions;
+
+    internal class ConnectionDataBase : Dictionary<LaneEnd, LaneConnectionData[]> {
+        internal bool IsConnectedTo(uint sourceLaneId, uint targetLaneId, ushort nodeId) =>
+            IsConnectedTo(sourceLaneId, targetLaneId, sourceLaneId.ToLane().IsStartNode(nodeId));
+
+        /// <param name="sourceStartNode">start node for the segment of the source lane</param>
+        internal bool IsConnectedTo(uint sourceLaneId, uint targetLaneId, bool sourceStartNode) {
+            var key = new LaneEnd(sourceLaneId, sourceStartNode);
+            if (this.TryGetValue(key, out var targets)) {
+                int n = targets.Length;
+                for (int i = 0; i < n; ++i) {
+                    if (targets[i].Enabled && targets[i].LaneId == targetLaneId) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Adds a connection from source to target lane at the give node
+        /// </summary>
+        internal void ConnectTo(uint sourceLaneId, uint targetLaneId, ushort nodeId) {
+            AddConnection(sourceLaneId, targetLaneId, nodeId, true); //forward
+            AddConnection(targetLaneId, sourceLaneId, nodeId, false); //backward
+        }
+
+        /// <summary>
+        /// makes a uni-directional connection
+        /// </summary>
+        /// <param name="enable"><c>true</c> for forward/bi-directional connection,
+        /// <c>false</c> for backward connection</param>
+        private void AddConnection(uint sourceLaneId, uint targetLaneId, ushort nodeId, bool enable) {
+            var key = new LaneEnd(sourceLaneId, nodeId);
+            {
+                if (this.TryGetValue(key, out var targets)) {
+                    int n = targets.Length;
+                    for (int i = 0; i < n; ++i) {
+                        if (targets[i].LaneId == targetLaneId) {
+                            // a uni-directional connection already exist
+                            targets[i].Enabled |= enable;
+                            return;
+                        }
+                    }
+                }
+            }
+
+            {
+                // no connection exists yet. create new connection.
+                var newConnection = new LaneConnectionData(targetLaneId, enable);
+                if (this.TryGetValue(key, out var targets)) {
+                    this[key] = targets.Append(newConnection);
+                } else {
+                    this[key] = new[] { newConnection };
+                }
+            }
```

Comment:
Do we need to do 2x `TryGetValue()`? Would something like this work...?

```csharp
var key = new LaneEnd(sourceLaneId, nodeId);

bool hasTargets = this.TryGetValue(key, out var targets);

if (hasTargets) {
    // ...
}

var newConnection = new LaneConnectionData(targetLaneId, enable);
this[key] = hasTargets
    ? targets.Append(newConnection)
    : new[] { newConnection };
```

Reply:
good idea! I can get rid of those extra blocks too :)